### ISSUE TITLE
Clarification needed

### ISSUE THREAD

**Issue by samsfisher** [forrert/mongoose-model-security#2]
How does this approach get the current user role? An example of implementation will be helpful. i Can see how the policy works but in terms of passing the req.user object to the model-security its a bit confusing.

**Comment by forrert** [comment]
This is indeed the trickiest part of it all. Have a look at [this section of the readme](https://github.com/forrert/mongoose-model-security#parameters). Model providers let you add values to the parameters argument in rule functions. You'll want to add a model provider, that returns an object holding the user of the current request. Getting access the current user is a little bit tricky unless you're working directly within a middleware function (due to the event based model of node.js). Using [the domain API](https://nodejs.org/api/domain.html) it is possible to hold on to the user object for the duration of a request.

I'll see if I can post some code to accomplish that.

**Comment by samsfisher** [comment]
@forrert 

Thanks for the clarification. You are right, that is the most trickiest part of all. I am concocting different plugins to do that and still not successful. if you can provide some sample code to accomplish that it will be of great help.

Thoughtful plugin though! ++ ;)

**Comment by forrert** [comment]
Glad you like the plugin. I'd love to hear more about how you use it!

Just a note on my previous comment: I originally tried to use [continuation-local-storage](https://www.npmjs.com/package/continuation-local-storage) to save the user object to the context of a request. Unfortunately this didn't work, I believe because some lower level library contained in mongoose reset the domain within certain operations. I ended up writing my own simplified version with a few work arounds. Now this was about half a year ago and things might have changed since. However you might be able to find a solution using the mentioned module (at least as a prototype).

**Comment by forrert** [comment]
Assuming you're working with express, here's a bit of code to get access to the user object:

``` js
// requestdomain.js
var domain = require('domain');

module.exports = {
    put: function(key, value) {
        if (domain.active && domain.active.context) {
            domain.active.context[key] = value;
        } else {
            throw new Error('No active domain at this point!');
        }
    },

    get: function(key) {
        if (domain.active && domain.active.context) {
            return domain.active.context[key];
        } else {
            return undefined;
        }
    }
};
```

``` js
// server.js
var app = ... // initialize express app
var security = ... // initialize security

var requestDomain = require('./requestdomain.js');

// create the model provider
security.addModelProvider(function() {
    var user = requestDomain.get('user');
    return {
        user: user ? user : null
    };
});

// register a route (assuming the user object is part of the request object)
app.use(function(req, res, next) {
    if (req.user) {
        requestDomain.put('user', req.user);
    } else {
        requestDomain.put('user', null);
    }
    next();
});

// build policy (assuming a 'Organisation' model, with a 'administrator' field)
security.buildPolicy('Organisation').
    update(function(parameters) {
        var userId = parameters.user ? parameters.user.id : null;
        if (userId) {
            return {administrator: userId};
        } else {
            return false;
        }
    });
```

Hope this helps!

**Comment by samsfisher** [comment]
I am trying to implement the above code but i dont seem to have domain.active set. Do we create a domain for every request? 

i get the error "No active domain at this point!"

**Comment by forrert** [comment]
I missed posting the part, where the domain is created in a middleware:

``` js
function middleware(req, res, next) {
    var reqDomain = domain.create();
    reqDomain.context = {};
    reqDomain.add(req);
    reqDomain.add(res);

    res.on('close', function() {
        reqDomain.exit();
    });

    res.on('finish', function() {
        reqDomain.exit();
    });

    reqDomain.on('error', function() {
        // just make sure the domain is exited
        reqDomain.exit();
    });

    reqDomain.run(function() {
        next();
    });
}
```

**Comment by samsfisher** [comment]
thanks

so the middleware code needs to go in requestdomain.js and the middleware is initialized in app.js??

**Comment by forrert** [comment]
Correct. Make sure you add the middleware early to your express app (possibly as your very first middleware, since you'd want a domain created for all requests right as they come in). Remember that some 3rd party libraries might mess with the domain created and remove it all together... You'd need to work around that manually somehow. I believe that particularly modules using add ons (linking C/C++ libraries into node) are likely to cause problems.
